ipn/ipnlocal/state_test.go

```diff
@@ -0,0 +1,778 @@
+	store := new(ipn.MemoryStore)
+	e, err := wgengine.NewFakeUserspaceEngine(logf, 0)
+	if err != nil {
+		t.Fatalf("NewFakeUserspaceEngine: %v", err)
+	}
+
+	cc := newMockControl()
+	ccGen := func(opts controlclient.Options) (controlclient.Client, error) {
+		cc.mu.Lock()
+		cc.opts = opts
+		cc.logf = opts.Logf
+		cc.authBlocked = true
+		cc.mu.Unlock()
+
+		cc.called("New")
+		return cc, nil
+	}
+	b, err := NewLocalBackendWithClientGen(logf, "logid", store, e, ccGen)
+	if err != nil {
+		t.Fatalf("NewLocalBackend: %v", err)
+	}
+
+	notifies := notifyThrottler{t: t}
+	notifies.expect(0)
+
+	b.SetNotifyCallback(func(n ipn.Notify) {
+		if n.State != nil ||
+			n.Prefs != nil ||
+			n.BrowseToURL != nil ||
+			n.LoginFinished != nil {
+			logf("\n%v\n\n", n)
+			notifies.put(n)
+		} else {
+			logf("\n(ignored) %v\n\n", n)
+		}
+	})
+
+	// Check that it hasn't called us right away.
+	// The state machine should be idle until we call Start().
+	assert.Equal(cc.getCalls(), []string{})
+
+	// Start the state machine.
+	// Since !WantRunning by default, it'll create a controlclient,
+	// but not ask it to do anything yet.
+	notifies.expect(2)
+	assert.Nil(b.Start(ipn.Options{
+		StateKey: ipn.GlobalDaemonStateKey,
+	}))
+	{
+		// BUG: strictly, it should pause, not unpause, here, since !WantRunning.
+		assert.Equal([]string{"New", "unpause"}, cc.getCalls())
+
+		nn := notifies.drain(2)
+		assert.Equal([]string{}, cc.getCalls())
+		assert.NotNil(nn[0].Prefs)
+		assert.NotNil(nn[1].State)
+		prefs := *nn[0].Prefs
+		assert.Equal(false, prefs.WantRunning)
+		// BUG: there should be a NoState first, to tell the GUI to show "Loading..."
+		assert.Equal(ipn.NeedsLogin, *nn[1].State)
+		assert.Equal(ipn.NeedsLogin, b.State())
+	}
+
+	// Restart the state machine.
+	// It's designed to handle frontends coming and going sporadically.
+	// Make the sure the restart not only works, but generates the same
+	// events as the first time, so UIs always know what to expect.
+	t.Logf("\n\nStart2")
+	notifies.expect(2)
+	assert.Nil(b.Start(ipn.Options{
+		StateKey: ipn.GlobalDaemonStateKey,
+	}))
+	{
+		// BUG: strictly, it should pause, not unpause, here, since !WantRunning.
+		assert.Equal([]string{"Shutdown", "New", "unpause"}, cc.getCalls())
+
+		nn := notifies.drain(2)
+		assert.Equal([]string{}, cc.getCalls())
+		assert.NotNil(nn[0].Prefs)
+		assert.NotNil(nn[1].State)
+		prefs := *nn[0].Prefs
+		assert.Equal(false, prefs.WantRunning)
+		// BUG: there should be a NoState first, to tell the GUI to show "Loading..."
+		assert.Equal(ipn.NeedsLogin, *nn[1].State)
+		assert.Equal(ipn.NeedsLogin, b.State())
+	}
+
+	// Start non-interactive login with no token.
+	// This will ask controlclient to start its own Login() process,
+	// then wait for us to respond.
+	t.Logf("\n\nLogin (noninteractive)")
+	notifies.expect(0)
+	b.Login(nil)
+	{
+		assert.Equal(cc.getCalls(), []string{"Login"})
+		notifies.drain(0)
+		// BUG: this should immediately set WantRunning to true.
```

Comment:
This leads to an undecided clash that's bitten us in the past, and continues to be inconsistent in our frontends: in some places, ipn.Prefs is treated as write-only user knobs, which only the user gets to alter and ipn only gets to respond to. In other places, we treat it as a general state store/scratchpad. And others are both.

It would be nice to rationalize this and document how we want this to work, so we stop drifting.

Reply:
Yup. I don't know how to solve that *whole* problem. But the most immediate problem in this case is that I would like to treat most of Prefs as "this is a recording of what the user would like things to be like." Mostly that originates in the UI, but sometimes the backend needs to update it.

So for the current example, when the user calls Login, I think it's reasonable to infer that the user wants WantRunning=true, so we just do that for them. Otherwise, the correct thing to do would be to login and then stay in disconnected mode and wait for them to click Connected (or worse: override this behaviour in every variant of the UI!), which is much worse.

However, that desirable behaviour (remember what the user wants) is not what the state machine currently does. Instead, it sets WantRunning=true after getting a login-related notification back from the control server. That makes little sense, and suggests there are probably even deeper weird edge cases hidden inside. We can fix this part at least.

Reply:
clarified comment.